torch/csrc/jit/codegen/cuda/grouped_reduction.cpp

```diff
@@ -0,0 +1,225 @@
+        output_tv->nDims() == num_dims,
+        "Invalid grouped reduction due to mismatched number of dimensions. "
+        "Expected: ",
+        num_dims,
+        ". Detected: ",
+        output_tv->nDims(),
+        ". Invalid output tensor: ",
+        output_tv->toString());
+
+    auto dep_other_input = getDependentVal(inputs, output_tv);
+    TORCH_INTERNAL_ASSERT(
+        dep_other_input == nullptr,
+        "Invalid dependency detected. ",
+        dep_other_input->toString(),
+        " depends on ",
+        output_tv->toString());
+
+    if (ref_tv == output_tv) {
+      continue;
+    }
+
+    for (const auto i : c10::irange(num_root_dims)) {
+      auto ref_id = ref_domain.at(i);
+      auto output_id = output_domain.at(i);
+      // If an IterDomain is broadcast, require the other
+      // corresponding IterDomains are also broadcast. This may not be
+      // necessary but not completely certain.
+      TORCH_INTERNAL_ASSERT(
+          ref_id->isBroadcast() == output_id->isBroadcast(),
+          "Invalid grouped reduction due to mismatched broadcast root domains. ",
+          "Reference domain: ",
+          ref_id->toString(),
+          ". Mismatched domain: ",
+          output_id->toString(),
+          ". Invalid tensor: ",
+          output_tv->toString());
+      if (ref_id->isBroadcast()) {
+        continue;
+      }
+      TORCH_INTERNAL_ASSERT(
+          ref_id->isReduction() == output_id->isReduction(),
+          "Invalid grouped reduction due to mismatched reduction root domains. ",
+          "Reference domain: ",
+          ref_id->toString(),
+          ". Mismatched domain: ",
+          output_id->toString(),
+          ". Invalid tensor: ",
+          output_tv->toString());
+      TORCH_INTERNAL_ASSERT(
+          exact_map.areMapped(ref_id, output_id) || ref_id->sameAs(output_id),
+          "Invalid grouped reduction due to mismatched root domains. ",
+          "Reference domain: ",
+          ref_id->toString(),
+          ". Mismatched domain: ",
+          output_id->toString(),
+          ". Invalid tensor: ",
+          output_tv->toString());
+    }
+
+    TORCH_INTERNAL_ASSERT(
+        hasMatchingTransformations(ref_tv, output_tv),
+        "Invalid grouped reduction due to mismatched transformations. ",
+        "Reference tensor: ",
+        ref_tv->toString(),
+        ". Mismatched tensor: ",
+        output_tv->toString());
+
+    // Must have the same computeAt position
+    TORCH_INTERNAL_ASSERT(
+        output_tv->getComputeAtPosition() == ref_ca_pos,
+        "Invalid grouped reduction due to mismatched computeAt position. ",
+        "Reference tensor: ",
+        ref_tv->toString(),
+        ". Mismatched tensor: ",
+        output_tv->toString());
+
+    max_producer_pos =
+        std::max(max_producer_pos, output_tv->getMaxProducerPosition());
+  }
+
+  return max_producer_pos;
+}
+
+} // namespace
+
+void groupReductions(const std::vector<TensorView*>& reduction_outputs) {
```

Comment:
couldn't we automatically group in pairs for however many were sent? I get that it won't today be one function, but seems like it should be fine without throwing an error. If there's odd number I just wouldn't group the last one with anything.

Reply:
There's no constraint on the number of reductions at this point. Currently there's only two-input version of grouped grid reduction, so if a grouped reduction becomes a grid grouped reduction, it would be an error if there's more than two reductions.

Reply:
I just meant if we get to lowering with grid reduction, we could just group it into multiple `gridReduction2`s.

Reply:
This isn't really critical, just a suggestion for the future if we care to.

Reply:
Oh, I see. I was thinking about creating a variadic argument version of grid reduction, but that's another option.